lib/codecs/src/encoding/format/syslog.rs

```diff
@@ -0,0 +1,577 @@
+use bytes::{BufMut, BytesMut};
+use tokio_util::codec::Encoder;
+use vector_core::{config::DataType, event::{Event, LogEvent}, schema};
+use chrono::{DateTime, SecondsFormat, SubsecRound, Utc};
+use vrl::value::{ObjectMap, Value};
+use vector_config::configurable_component;
+
+use std::collections::HashMap;
+
+// All of this block is to support the Facility + Severity enums with convenience of string or ordinal config value:
+use std::str::FromStr;
+use strum::{FromRepr, EnumString, VariantNames};
+// `akin` macro for DRY impl to share with both enums due to lack of a `FromRepr` trait:
+use akin::akin;
+// Custom deserialization with serde needed:
+use serde::{Deserialize, Deserializer, de::Error};
+use serde_aux::field_attributes::deserialize_number_from_string;
+use snafu::{Snafu, OptionExt};
+
+/// Config used to build a `SyslogSerializer`.
+#[configurable_component]
+#[derive(Clone, Debug, Default)]
+#[serde(default)]
+pub struct SyslogSerializerConfig {
+    /// Options for the Syslog serializer.
+    pub syslog: SyslogSerializerOptions
+}
+
+impl SyslogSerializerConfig {
+    /// Build the `SyslogSerializer` from this configuration.
+    pub fn build(&self) -> SyslogSerializer {
+        SyslogSerializer::new(&self)
+    }
+
+    /// The data type of events that are accepted by `SyslogSerializer`.
+    pub fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    /// The schema required by the serializer.
+    pub fn schema_requirement(&self) -> schema::Requirement {
+        schema::Requirement::empty()
+    }
+}
+
+/// Syslog serializer options.
+#[configurable_component]
+#[derive(Clone, Debug, Default)]
+// Serde default makes all config keys optional.
+// Each field assigns either a fixed value, or field name (lookup field key to retrieve dynamic value per `LogEvent`).
+#[serde(default)]
+pub struct SyslogSerializerOptions {
+    /// RFC
+    rfc: SyslogRFC,
+    /// Facility
+    #[serde(deserialize_with = "Facility::deserialize")]
+    facility: Facility,
+    /// Severity
+    #[serde(deserialize_with = "Severity::deserialize")]
+    severity: Severity,
+
+    /// App Name
+    app_name: Option<String>,
+    /// Proc ID
+    proc_id: Option<String>,
+    /// Msg ID
+    msg_id: Option<String>,
+
+    /// Payload key
+    payload_key: String,
```

Comment:
| Remove this option. This should always be the field with the meaning message.

I would disagree with this suggestion as it reduces the flexibility of using the protocol.  This implies you always want the "message" portion of the payload to be what vector determines is the "message" and not for instance how an administrator may have transformed it in a prior pipeline to be something else.  The default if not specified IMO should be "message" but allow the flexibility; there is no reason not to

Reply:
> This implies you always want the "message" portion of the payload to be what vector determines is the "message" and not for instance how an administrator may have transformed it in a prior pipeline to be something else.

Could you elaborate why you'd have introduced a difference in semantics? Please document a valid use-case for why you'd want such functionality so we know it's not an XY problem.

When would you transform the log message to not represent the message, yet provide it as input to an encoder to but require a different key for the actual message content?

---

I am familiar with the original PR use with `add_source` to compose metadata related to k8s with the actual message. You could do that with VRL, or for the metadata leverage structured data?

Reply:
> Could you elaborate why you'd have introduced a difference in semantics? Please document a valid use-case for why you'd want such functionality so we know it's not an XY problem.

Vector is configurable as a transformation engine.  Logs coming in do not need to equal logs going out.  Consider almost any  case where my desired payload is a subset of the original message, a value in the message which was a JSON string.  In a prior transform, I could parse the message and now it is available to me as key/value pairs, where I only care about some portion of the message.

By removing this option does it mean in some prior transform do I always have to modify the event and set the key "message" to the desired outcome?

Reply:
> By removing this option does it mean in some prior transform do I always have to modify the event and set the key "message" to the desired outcome? Maybe I'm missing where the encoding ends and transforms and sinks begin 🤷‍♂️

Encoding is `A => B`, given input A produce output B (syslog formatted log in this case).

Perhaps I should phrase this a bit differently?:
- If Vector supports encoding your input to other existing log encoders the way you want, you should expect to get the same with this syslog encoder.
- If it does not, then there would need to be a very valid reason why this encoder is an exception vs what you'd otherwise need to do with any other encoder already supported by Vector.

As such, you should find that what you're requesting isn't specific to this encoder implemented via the PR, but something one would expect to be encoder agnostic.

That would be a separate feature request to raise, which may be rejected on the basis that VRL would be advised for your use-case to pre-process your input for the encoder instead.

---

I'm not an active Vector user myself yet, but from what I do understand `message` is an internal convention and the encoders rely on that to be semantically correct as input, thus you'd be expected to use existing Vector features to handle that prior to encoding.

@StephenWakely will chime in if I'm mistaken, and you're also welcome to correct me too.